docs: add change doc to describe key namespaces, names, provider IDs, and provider info (#135)
awslabs/aws-encryption-sdk-specification: changes/2020-06-09_wrapping-key-identifiers/change.md

## Summary

We have referenced several terms throughout various specification documents
that are never defined:
"key namespace",
"key name",
"provider ID",
and "provider info".
This change defines each of these terms
and describes their relationships to each other
and the keyrings they represent.

## Out of Scope

- Whether, or how, each keyring exposes any of these values at runtime is out of scope.

- Specific values for these terms for any specific keyring is out of scope.

## Motivation

Various specification documents reference
"key namespace",
"key name",
"provider ID",
and "provider info",
but we never define what these terms mean
or how they relate to each other.

In order to ensure the specification documents accurately describe keyring behavior,
we need to define all terms that they use.

## Drawbacks

This change SHOULD NOT introduce any drawbacks.
It is identifying and describing
terms that already exist
and the existing relationship between them.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

This change SHOULD NOT have any operational implications.

## Guide-level Explanation

> Key namespace and key name are configuration values that MAY determine the behavior of a keyring.

Reviewer: There is something about this MAY that makes me a little uncomfortable.
Like a keyring that does not delegate, may not even have these values.
Below you talk about how the values SHOULD be equal.
Perhaps here is a place to express the relationship more concretely?

I understand that
the "key namespace" and the "key name" as "for the keyring"
and the "provider ID and "provider info" as "for the encrypted data key".

Provider ID and provider info are values that identify the keyring configuration
that can fulfill the keyring's decryption contract.
The keyring attaches these values to the encrypted data key ciphertext.

## Reference-level Explanation

"Key namespace", "key name", "provider ID", "provider info"
are all concepts that identify a wrapping key.

### key namespace

A configuration value for a keyring
that identifies the grouping or categorization
for the wrapping keys that keyring can access.

### key name

A configuration value for a keyring
that identifies a single wrapping key
within a key namespace.

### provider ID

An output value returned by a keyring on encrypt
as part of an encrypted data key structure
that identifies the grouping or categorization
for the keyring that can fulfill this keyring's decryption contract.

The provider ID SHOULD be equal to the key namespace.

### provider info

An output value returned by a keyring on encrypt
as part of an encrypted data key structure
that provides necessary information for a keyring
to fulfill this keyring's decryption contract.

> The provider info SHOULD be equal to the key name.

Reviewer 2: Note: "provider info", and "key name" are defined with slightly different types.

"provider info" is defined as "usually a UTF-8 encoded value", where as "key name" is defined specifically as a "UTF-8 encoded value" in both raw keyrings.

Reviewer: This is true, and ideally we would have some _why_ written down about why "provider info" is bytes and not a string

> One example of a keyring where the key name and provider info can differ

Reviewer: It the Raw AES keyring another good example?

Author: It's an example...I don't know that it's a *good* example. :/

I'd rather not call attention to it shoving extra stuff into the provider info because that's a pattern that we do not want to encourage in the future.

Reviewer: totally fair

in this document are to be interpreted as described in
[RFC 2119](https://tools.ietf.org/html/rfc2119).

## Summary

We have referenced several terms throughout various specification documents
that are never defined:
"key namespace",
"key name",
"provider ID",
and "provider info".
This change defines each of these terms
and describes their relationships to each other
and the keyrings they represent.

## Out of Scope

- Whether, or how, each keyring exposes any of these values at runtime is out of scope.

- Specific values for these terms for any specific keyring is out of scope.

## Motivation

Various specification documents reference
"key namespace",
"key name",
"provider ID",
and "provider info",
but we never define what these terms mean
or how they relate to each other.

In order to ensure the specification documents accurately describe keyring behavior,
we need to define all terms that they use.

## Drawbacks

This change SHOULD NOT introduce any drawbacks.
It is identifying and describing
terms that already exist
and the existing relationship between them.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

> If the keyring chooses to write a provider info value that does not equal its key name,

Reviewer 3: nit: this doesn't appear to be an operational implication for defining the terms "key name" and "provider info", this is just operational implications of how these concepts as they currently exist (which we aren't changing).

[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"

# Clarify Wrapping Key Identifiers

## Affected Features

This serves as a reference of all features that this change affects.

| Feature                                                                                             |
| --------------------------------------------------------------------------------------------------- |
| [Encrypted Data Key](../../framework/structures.md#encrypted-data-key)                              |
| [AWS KMS Keyring](../../framework/kms-keyring.md)                                                   |
| [Raw AES Keyring](../../framework/raw-aes-keyring.md)                                               |
| [Raw RSA Keyring](../../framework/raw-rsa-keyring.md)                                               |

> | [Keyring Decryption Contract](https://github.com/awslabs/aws-encryption-sdk-specification/pull/131) |

Reviewer 2: @acioc had the idea of making these permanent links to the specific commit before the change.

Reviewer 4: I think there's some ongoing discussion of this on https://github.com/awslabs/aws-encryption-sdk-specification/issues/137

and describes their relationships to each other
and the keyrings they represent.

## Out of Scope

- Whether, or how, each keyring exposes any of these values at runtime is out of scope.

- Specific values for these terms for any specific keyring is out of scope.

## Motivation

Various specification documents reference
"key namespace",
"key name",
"key provider ID",
and "key provider info",
but we never define what these terms mean
or how they relate to each other.

In order to ensure the specification documents accurately describe keyring behavior,
we need to define all terms that they use.

## Drawbacks

This change SHOULD NOT introduce any drawbacks.
It is identifying and describing
terms that already exist
and the existing relationship between them.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

This change SHOULD NOT have any operational implications.

## Guide-level Explanation

Key namespace and key name are configuration values that determine the behavior of a keyring.

Provider ID and key provider info are values that identify the keyring configuration
that can fulfill the keyring's decryption contract.

> The keyring attaches these values to the encrypted data key ciphertext.

Reviewer 2: ```suggestion
The keyring attaches these values to a data key ciphertext to form an encrypted data key.
```

## Reference-level Explanation

"Key namespace", "key name", "provider ID", and "provider info"
are all concepts that identify a wrapping key.

### key namespace

A configuration value for a keyring
that identifies the grouping or categorization
for the wrapping keys that keyring can access.

The key namespace MUST be a string value.

### key name

A configuration value for a keyring
that identifies a single wrapping key
within a key namespace.

The key name MUST be a string value.

> ### provider ID

Reviewer 5: Is this implicitly a "key provider ID" and should we qualify it as such, or are you intentionally trying to disambiguate it from the "key provider ID" in an EDK?

Author: I forgot we had that named with "key". I'll update and link to the structures value.

An output value returned by a keyring on encrypt
as part of an encrypted data key structure
that identifies the grouping or categorization

> for the keyring that can fulfill this keyring's decryption contract.

Reviewer: ```suggestion
for a keyring to fulfill this its decryption contract.
```

From #131, I'm still wondering about where
the decryption contract is.
I _think_ that the contract **is** the encrypted data key?
But then what do we say about keyrings,
that that can unwrap the encrypted data key
but chose not to?

Author: > But then what do we say about keyrings,
that that can unwrap the encrypted data key
but chose not to?

In that scenario I would say that is an example of a keyring being "unwilling" (as opposed to "unable") to fulfill the decryption contract.
